arbor/util/dylib.cpp

```diff
@@ -0,0 +1,59 @@
+#include <fstream>
+#include <string>
+
+#include <dlfcn.h>
+
+#include <arbor/arbexcept.hpp>
+
+#include "util/dylib.hpp"
+#include "util/strprintf.hpp"
+
+namespace arb {
+namespace util {
+
+struct dl_handle {
+    void* dl = nullptr;
+};
+
+dl_handle dl_open(const std::string& fn) {
+    try {
+        std::ifstream fd{fn.c_str()};
+        if(!fd.good()) throw file_not_found_error{fn};
+    } catch(...) {
+        throw file_not_found_error{fn};
+    }
+    // Call once to clear errors not caused by us
+    dlerror();
+    auto result = dlopen(fn.c_str(), RTLD_LAZY);
+    // dlopen fails by returning NULL
+    if (nullptr == result) {
+        auto error = dlerror();
+        throw dl_error{util::pprintf("[POSIX] dl_open failed with: {}", error)};
+    }
+    return {result};
+}
+
+void dl_close(dl_handle& handle) {
+    dlclose(handle.dl);
+    handle.dl = nullptr;
+}
+
+namespace impl{
+void* dl_get_symbol(const std::string& fn, const std::string& symbol) {
```

Comment:
What's the reason for the change from 
```cxx
void* dl_get_symbol(void* handel, const std::string& symbol)
```
? Now, fetching a symbol will go through `dlopen` everytime.

Reply:
The library was only being used to return a symbol from a library, for which the `handle` type was not required.

This way the handle type is not required in the public interface. It could be reintroduced later, if we ever have more elaborate use cases for dynamic libraries.

Reply:
Good point.